### ISSUE TITLE
Compact printing in cut

### ISSUE THREAD

**Issue by bkamins** [JuliaData/CategoricalArrays.jl#381]
Currently we have:
```julia
julia> cut(1:10, 3)
10-element CategoricalArray{String,1,UInt32}:
 "Q1: [1.0, 4.0)"
 "Q1: [1.0, 4.0)"
 "Q1: [1.0, 4.0)"
 "Q2: [4.0, 6.999999999999999)"
 "Q2: [4.0, 6.999999999999999)"
 "Q2: [4.0, 6.999999999999999)"
 "Q3: [6.999999999999999, 10.0]"
 "Q3: [6.999999999999999, 10.0]"
 "Q3: [6.999999999999999, 10.0]"
 "Q3: [6.999999999999999, 10.0]"
```
which is not nice. It would be better to use compact printing.

Though we should make sure to correctly do this case:
```julia
julia> cut(1:10^-12:1+10^-11, 3)
11-element CategoricalArray{String,1,UInt32}:
 "Q1: [1.0, 1.0000000000033333)"
 "Q1: [1.0, 1.0000000000033333)"
 "Q1: [1.0, 1.0000000000033333)"
 "Q1: [1.0, 1.0000000000033333)"
 "Q2: [1.0000000000033333, 1.0000000000066667)"
 "Q2: [1.0000000000033333, 1.0000000000066667)"
 "Q2: [1.0000000000033333, 1.0000000000066667)"
 "Q3: [1.0000000000066667, 1.00000000001]"
 "Q3: [1.0000000000066667, 1.00000000001]"
 "Q3: [1.0000000000066667, 1.00000000001]"
 "Q3: [1.0000000000066667, 1.00000000001]"
```

**Comment by andreasnoack** [comment]
I just hit a similar case hwere the many digits made plotting labels look ugly. What about supporting rounding in the `cut(array, ngroups)` method instead of just changing the printing? I guess it might be confusing if the string here don't reflect the actual cuts.

**Comment by nalimilan** [comment]
This is tricky. Rounding could easily change radically the size of the groups if values are very close. AFAIK other packages don't do that. What R's `cut` does is that it tries with 3 digits by default, and if some breaks end up represented the same it increases the number of digits up to 12. This can be tweaked via the `dig.lab` argument. This sounds reasonable to me at least for quantiles, as you probably don't care about the exact value in that case. OTOH it could be more surprising when specifying breaks manually.

At the very least we could add that `dig.lab` argument.

**Comment by andreasnoack** [comment]
The `Qx` label doesn't specify the probability of the quantile anyway. Actually, I'd prefer a label without `Qx` part. I'm mostly interested in the interval information. It is easy to incorrectly think that `Q1` is the first quartile.

**Comment by nalimilan** [comment]
I added `Qx` because when `allowempty=true` some intervals may be identical (in the presence of many duplicates), but we can't have levels with equal names. But that's a corner case so we could stop doing that by default. Still, returning quantiles that are not real quantiles due to rounding would be confusing IMO.

**Comment by nalimilan** [comment]
This is related to a discussion we had in 2020 at https://github.com/JuliaData/CategoricalArrays.jl/pull/245. @bkamins Do you have an opinion?

**Comment by nalimilan** [comment]
This is related to a discussion we had in 2020 at https://github.com/JuliaData/CategoricalArrays.jl/pull/245. @bkamins Do you have an opinion? I see several options:
- Keep adding `Qx`.
- Provide a formatter that adds `Qx` and advise using it when we throw an error because of duplicate intervals.
- Support an argument doing that.
- Automatically use that formatter by default when `allowempty=true` is passed.

**Comment by bkamins** [comment]
Given the discussion - I think adding an option is most flexible.

**Comment by nalimilan** [comment]
Can you elaborate why you prefer an argument over a formatter? I find it hard to decide which is best.

**Comment by andreasnoack** [comment]
After thinking a bit further, I see two separate but somewhat related issues here.

1. Which cuts to use
2. Which labels to attach

This issue is of course about 2. but now I think that 1. is actually what makes 2. an bigger issue. At least to me. In my use cases, I'd prefer cuts at the values in the input vector, i.e. without linear interpolation. So R-1 in https://en.wikipedia.org/wiki/Quantile#Estimating_quantiles_from_a_sample. Typically, real data doesn't have the many decimals so using the R-1 definition would avoid very long representations of the cuts and therefore 2. is no longer as relevant (to me). Is that something you'd be willing to consider here?

> I added `Qx` because when `allowempty=true` some intervals may be identical (in the presence of many duplicates), but we can't have levels with equal names.

I guess you have thought this through many years ago but having non-unique intervals seems unfortunate to me. I can see that something would have to go, though, but I think I'd prefer to change the definition such that `cut` produces at most `n` cuts.

Btw. found https://github.com/JuliaStats/Statistics.jl/pull/184 when playing around with this issue.

**Comment by nalimilan** [comment]
> This issue is of course about 2. but now I think that 1. is actually what makes 2. an bigger issue. At least to me. In my use cases, I'd prefer cuts at the values in the input vector, i.e. without linear interpolation. So R-1 in https://en.wikipedia.org/wiki/Quantile#Estimating_quantiles_from_a_sample. Typically, real data doesn't have the many decimals so using the R-1 definition would avoid very long representations of the cuts and therefore 2. is no longer as relevant (to me). Is that something you'd be willing to consider here?

That's an interesting idea. On the one hand we probably don't care a lot about the exact cutpoints as long as the resulting quantiles are identical, and using those from actual data makes sense. On the other one, getting results slightly different from `quantile` (which doesn't even support R-1 currently) is a bit annoying. And it wouldn't completely solve the problem, as even simple real data can contain many decimals (log income, average income per household member...), so we would need a solution for compact printing anyway. So I'm hesitant...

In terms of implementation, the cleanest solution would be to change `quantile` to support R-1 first, but it won't be available on old Julia versions that we want to support (like 1.6) to avoid creating a big pre-1.0 vs post-1.0 CategoricalArrays divide. Instead we would have to take the sorted array returned by `quantile` and find the data points corresponding to quantiles.

> I guess you have thought this through many years ago but having non-unique intervals seems unfortunate to me. I can see that something would have to go, though, but I think I'd prefer to change the definition such that `cut` produces at most `n` cuts.

This is opt-in so we can offer several options. The idea behind the current `allowempty=true` behavior is that sometimes you really need to get the expected number of groups so that e.g. your tables and models are consistent across countries (this is something that happened to me in practice). I guess in other cases you'd like as many groups as possible up to `ngroups` so we could also support that (it's easy). Either we can add an argument like `allowfewer=true`, or maybe better use a single argument like `duplicatebreaks=:error/:empty/:skip`?

**Comment by andreasnoack** [comment]
> And it wouldn't completely solve the problem, as even simple real data can contain many decimals (log income, average income per household member...), so we would need a solution for compact printing anyway. So I'm hesitant...

That is true but my guess is that it would be sufficient in many cases and it would have the advantage that the user could manually round that quantity before passing it to `cut` if needed.

> In terms of implementation, the cleanest solution would be to change quantile to support R-1 first, but it won't be available on old Julia versions that we want to support (like 1.6) to avoid creating a big pre-1.0 vs post-1.0 CategoricalArrays divide.

I agree that it would be nice to have in `quantile`. The actual algorithm would be very easy to implement. I guess the main challenge would figure out the interface since the current interface is based on `alpha` and `beta`. Maybe we could have a separate `method`/`definition` keyword where the user is only allowed to specify either `definition` or `alpha`/`beta`. I will open an issue if `Statistics` to discuss this.

**Comment by bkamins** [comment]
And what if assuming we do not touch the cut-points we provide a modification of labels that chooses the shortest label that does not change the split? In other words assuming we have an initial label `L` and it falls between values `x` and `y` in data, i.e. we have `x <= L <= y` and there are no data points between `x` and `y` we pick a corrected label `L'` such that `x <= L' <= y` and `L'` has as short representation as possible?

**Comment by nalimilan** [comment]
Yeah I had the same thought. Maybe we could take R-1 quantiles as starting points (which give `y` in your notation AFAICT), and call `floor(y, ndigits)` on them? If the resulting value is still greater than the next lower value (`x`), use it as a cutpoint, and otherwise increase `ndigits` and start over?

I'm not sure how the implementation would work though (assuming we add R-1 quantiles to `quantile`). Two ideas:
- Call `quantile!` on a copy of the input, then `partialsort!` to ensure values in positions right before quantiles are sorted, and extract them.
- After calling `quantile` to get R-1 quantiles, iterate over the input and collect for each quantile the greatest value which is lower than it.

Not sure which one is faster. The second one at least will continue working automatically if we support weighted quantiles at some point.

**Comment by bkamins** [comment]
If we use the R-1 quantiles then indeed we have `y`. Then we need to do one pass over data to collect for each quantile the greatest value lower then it (`x`). Then we can check up to which digit after the decimal point `y` and `x` are the same and we need use one more digit than that.

For example, if `x = 1234.5678` and `y=1234.5681` we know that we have to use `1234.568` as a cut point since the two numbers are the same with just two decimal digits.

The tricky part is when scientific notation gets introduced, but I think this also can be sorted out. Here is an example hard case:

```
julia> x
30-element Vector{Float64}:
 8.85718832925723e-7
 8.572446994052413e-7
 1.40217695121027e-7
 8.966449714804087e-7
 3.070384341319471e-7
 1.8520709563325888e-7
 2.1409955523249603e-7
 5.630461710066611e-7
 6.781422109070843e-7
 4.776113711396994e-7
 0.2538909094146984
 0.5249665525921473
 0.8321957380046366
 0.9648282851978118
 0.36084175275805797
 0.7851054639425253
 0.6875195857202754
 0.614940093507575
 0.6224944997292978
 0.6055683461790675
 5.349085340927365e11
 1.3471583229449602e11
 6.538893396835975e11
 4.826316844547661e11
 8.803607035550856e11
 1.8174694671397316e10
 1.6709745443719125e11
 3.2050577954311835e11
 1.6134999167460663e11
 7.396308745225059e11

julia> cut(x, 10)
30-element CategoricalArray{String,1,UInt32}:
 "Q3: [6.551230029269997e-7, 8.933671299140029e-7)"
 "Q3: [6.551230029269997e-7, 8.933671299140029e-7)"
 "Q1: [1.40217695121027e-7, 2.9774454624200196e-7)"
 "Q4: [8.933671299140029e-7, 0.4593166326585115)"
 "Q2: [2.9774454624200196e-7, 6.551230029269997e-7)"
 "Q1: [1.40217695121027e-7, 2.9774454624200196e-7)"
 "Q1: [1.40217695121027e-7, 2.9774454624200196e-7)"
 "Q2: [2.9774454624200196e-7, 6.551230029269997e-7)"
 "Q3: [6.551230029269997e-7, 8.933671299140029e-7)"
 "Q2: [2.9774454624200196e-7, 6.551230029269997e-7)"
 "Q4: [8.933671299140029e-7, 0.4593166326585115)"
 "Q5: [0.4593166326585115, 0.6187172966184364)"
 "Q7: [0.8039415735673698, 5.313703595832701e10)"
 "Q7: [0.8039415735673698, 5.313703595832701e10)"
 "Q4: [8.933671299140029e-7, 0.4593166326585115)"
 "Q6: [0.6187172966184364, 0.8039415735673698)"
 "Q6: [0.6187172966184364, 0.8039415735673698)"
 "Q5: [0.4593166326585115, 0.6187172966184364)"
 "Q6: [0.6187172966184364, 0.8039415735673698)"
 "Q5: [0.4593166326585115, 0.6187172966184364)"
 "Q9: [1.977791194583766e11, 5.4680661465182275e11)"
 "Q8: [5.313703595832701e10, 1.977791194583766e11)"
 "Q10: [5.4680661465182275e11, 8.803607035550856e11]"
 "Q9: [1.977791194583766e11, 5.4680661465182275e11)"
 "Q10: [5.4680661465182275e11, 8.803607035550856e11]"
 "Q7: [0.8039415735673698, 5.313703595832701e10)"
 "Q8: [5.313703595832701e10, 1.977791194583766e11)"
 "Q9: [1.977791194583766e11, 5.4680661465182275e11)"
 "Q8: [5.313703595832701e10, 1.977791194583766e11)"
 "Q10: [5.4680661465182275e11, 8.803607035550856e11]"
```

**Comment by nalimilan** [comment]
Yeah handling both classic and scientific notation is the hardest. A possible trick would be to call `string(max(ceil(y, digits=ndigits), ceil(y, sigdigits=ndigits)))` (which AFAICT allows getting `ndigits` digits for both notations) and check whether it gives different values for consecutive cut points, and increase `ndigits` if not. That would avoid diving into representation details ourselves, which could be an advantage in case some custom numeric types use different representations from Base (does that exist currently?).

**Comment by nalimilan** [comment]
https://github.com/JuliaData/CategoricalArrays.jl/pull/422 implements a solution based on simply rounding breaks. As noted there, calling `floor` doesn't give great results, like with `floor(0.0003, sigdigits=4)` giving 0.0002999. This is so weird for users that in the end it seems better to round (this is what R does) even it it's less correct. Anyway even with `floor`, there's the risk that the break becomes lower than the highest value in the previous interval, and avoiding this requires making the behavior less predictable for users (see #422). Finding a mathematically correct solution runs the risk of over-engineering.
